### ISSUE TITLE
DiscordAPIError: 405: Method Not Allowed | Crashes While Bot is Invited to Any Server

### ISSUE THREAD

**Issue by cwerdna19** [codetheweb/muse#637]
**What's happening?**
Crashes immediately after inviting to a server.
Tested changing permission integer to 8 (Administrator) out of curiosity but no change.
Have invited it to two servers. One I am the owner of, the other I just have manage server permissions. Only invited the bot to the server I own as Administrator.
Each subsequent attempt to run the bot while it is in any guild will crash until the bot is kicked from all servers.
Appears to crash while processing permissions.

**Logs**

```
✔ Database migrations applied.
✔ Ready! Invite the bot with https://discordapp.com/oauth2/authorize?client_id=<censored>&scope=bot%20applications.commands&permissions=36700288
/root/muse/node_modules/discord.js/src/rest/RequestHandler.js:350
      throw new DiscordAPIError(data, res.status, request);
            ^

DiscordAPIError: 405: Method Not Allowed
    at RequestHandler.execute (/root/muse/node_modules/discord.js/src/rest/RequestHandler.js:350:13)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async RequestHandler.push (/root/muse/node_modules/discord.js/src/rest/RequestHandler.js:51:14)
    at async ApplicationCommandPermissionsManager.set (/root/muse/node_modules/discord.js/src/managers/ApplicationCommandPermissionsManager.js:186:18)
    at async updatePermissionsForGuild (file:///root/muse/src/utils/update-permissions-for-guild.ts:32:3)
    at async Client.guild_create_default (file:///root/muse/src/events/guild-create.ts:36:3) {
  method: 'put',
  path: '/applications/968996889099911238/guilds/450306155361861632/commands/permissions',
  code: 0,
  httpStatus: 405,
  requestData: {
    json: [
      {
        id: '968999097581322281',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097581322282',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097581322283',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097581322284',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097581322285',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097581322286',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097581322287',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097581322288',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097581322289',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097581322290',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097652637696',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097652637697',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097652637698',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097652637699',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097652637700',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097652637701',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097652637702',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      },
      {
        id: '968999097652637703',
        permissions: [
          { id: '148334523514486784', permission: true, type: 2 },
          { id: '450306155361861632', permission: false, type: 1 },
          { id: '148334523514486784', permission: true, type: 2 }
        ]
      }
    ],
    files: []
  }
}
file:///root/muse/node_modules/execa/lib/error.js:59
                error = new Error(message);
                        ^

Error: Command failed with exit code 1: esmo src/scripts/migrate-and-start.ts
    at makeError (file:///root/muse/node_modules/execa/lib/error.js:59:11)
    at handlePromise (file:///root/muse/node_modules/execa/index.js:119:26)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async file:///root/muse/src/scripts/run-with-database-url.ts:5:3 {
  shortMessage: 'Command failed with exit code 1: esmo src/scripts/migrate-and-start.ts',
  command: 'esmo src/scripts/migrate-and-start.ts',
  escapedCommand: 'esmo "src/scripts/migrate-and-start.ts"',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: undefined,
  stderr: undefined,
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}

```

**Screenshots**
N/A

**Additional context**
Fresh install of v1.9.0 branch.
Installed using command `npm i`.
Run using command `npm run start`.

**Runtime**
I'm running Muse:
- [x] Directly from the cloned repository
- [ ] Inside a Docker container
- [ ] Something else (please elaborate)

**Versions**
- Muse: 1.9.0
- Docker (if applicable): N/A
- OS: Ubuntu Server 16.04 LTS

**Comment by Makifun** [comment]
Looks like Discord updated some stuff yesterday, it also mentions changes to batch editing of permissions here:
[https://discord.com/developers/docs/change-log](https://discord.com/developers/docs/change-log)

Tried going back to older versions of the bot and got the same error.

**Comment by kkendzia** [comment]
Issue due to breaking change in discord API from April 27th:
https://discord.com/developers/docs/change-log

`Disabled the batch editing endpoint ([PUT /applications/{application.id}/guilds/{guild.id}/commands/permissions](https://discord.com/developers/docs/interactions/application-commands#batch-edit-application-command-permissions)).`

**Comment by joloman2** [comment]
same issue over here is there a fix for it ???

**Comment by kkendzia** [comment]
So I've taken 2 hours now and tried to understand the overall issues here.

@codetheweb 
I think you probably already figured it out but here are my "investigation" results.
The major problem is now, that you can't set application command permissions with a normal bot token anymore. So the authentication process of the bot has to be changed to oauth2 flow (shortend version https://discord.com/developers/docs/topics/oauth2#client-credentials-grant) - discord.js can handle the bearer token, but also we need a functionality to refresh the token from time to time. 
Also the batch processing of permission changes got deactivated, which is a small fix (1 req per command per guild instead of 1 request per guild for all commands). We have to take care about rate limit (50 req / second) when there are too much entries. 

I tried to implement something but my knowledge about js/ts is just too little to achieve something useful. Probably instead of taking the discord_token pass discord_client_Id and discord_client_Secret to the bot, which initially retrieves a token and  pass it to bot.ts. But didn't make something up for refreshing the token.
Hope this helps a bit. I will try to work further on my stuff to achieve something and learn how to write js/ts.

**Comment by codetheweb** [comment]
I'll try to take a look at this tomorrow.

PRs welcome if someone figures out a fix. :)

**Comment by kkendzia** [comment]
The necessary change in discordjs/rest got commited yesterday https://github.com/discordjs/discord.js/commit/679dcda9709376f37cc58a60f74d12d324d93e4e and haven't been released in a new version yet... So we have to wait I think.

**Comment by h45h74x** [comment]
Hi,

first of all: Thank you for the great project! Despite the current difficulties, I managed to add muse to my server and I'm really enjoying it! :D


I've tried to make sense of the sudden (breaking) changes Discord indroduced with their [API update](https://discord.com/developers/docs/change-log#updated-command-permissions). 

From reading through the discord.js reddit and several GitHub issues, it appears that with the recent update, bots are not supposed to update permissions anymore - users should. Permissions are now meant to be set by server itself, not by the bot.

To test this, I've disabled the `updatePermissionsForGuild` method in `muse/src/utils/update-permissions-for-guild.ts` with a `return;` statement (otherwise the bot will crash on invite) and tried to set the command permissions by myself. And indeed, it seems to work. I can manage the permissions for any Slash-Command the bot offers via `Server Settings > Integrations` in the Discord client.

There's only one thing: As muse seems to add the unconfigured commands to the server, you have to desync the command permissions, for example by adding a dummy role to the user section. Only after that desync, the commands show up in the server.


I recognize that the 'messing with permissions' approach isn't really a great solution and most certainly not how muse is to be used, but this could be a workaround until the underlying problem is fixed. [Here](https://i.imgur.com/u9372tN.png) is a screenshot on how I've configured my commands.

**Comment by kkendzia** [comment]
Hi @h45h74x 
I opend up a [PR](https://github.com/codetheweb/muse/pull/640), maybe you also can take a look into branch https://github.com/kkendzia/muse/tree/fix/command_permission.
With adding oauth2 capabilities the bot is able to set permissions.  Feel free to add comments!

**Comment by joloman2** [comment]
@kkendzia does that one one work for now till we get a official fix ???

**Comment by h45h74x** [comment]
> @kkendzia does that one one work for now till we get a official fix ???

@kkendzia's version is up and running on my server. You just have to fill out the new `DISCORD_CLIENT_ID` and `DISCORD_CLIENT_SECRET` variables and adapt the docker image to use `codetheweb/muse:pr-640` - then you're good to go :tada: 

![image](https://user-images.githubusercontent.com/27204033/166119336-ebbdf351-d55c-41ac-aa96-1b863308d2c1.png)

**Comment by joloman2** [comment]
okey so im also running it in  docker compose and this is what it need right 

`version: '3.4'

services:
  muse:
    image: codetheweb/muse:PR-640
    restart: always
    volumes:
      - ./muse:/data
    environment:
      - DISCORD_CLIENT_ID=
	  - DISCORD_CLIENT_SECRET=
      - YOUTUBE_API_KEY=
      - SPOTIFY_CLIENT_ID=
      - SPOTIFY_CLIENT_SECRET=`

**Comment by h45h74x** [comment]
> okey so im also running it in docker compose and this is what it need right
> 
> version: '3.4'
> 
> services: muse: image: codetheweb/muse:PR-640 restart: always volumes: - ./muse:/data environment: - DISCORD_CLIENT_ID= - DISCORD_CLIENT_SECRET= - YOUTUBE_API_KEY= - SPOTIFY_CLIENT_ID= - SPOTIFY_CLIENT_SECRET=

Looks good, just try it out

**Comment by joloman2** [comment]
> > okey so im also running it in docker compose and this is what it need right
> > version: '3.4'
> > services: muse: image: codetheweb/muse:PR-640 restart: always volumes: - ./muse:/data environment: - DISCORD_CLIENT_ID= - DISCORD_CLIENT_SECRET= - YOUTUBE_API_KEY= - SPOTIFY_CLIENT_ID= - SPOTIFY_CLIENT_SECRET=
> 
> Looks good, just try it out. I left my `DISCORD_TOKEN` variable in as well tho, I don't know if it's still needed.

okey where can i find the client id and secret ???

**Comment by h45h74x** [comment]
> > > okey so im also running it in docker compose and this is what it need right
> > > version: '3.4'
> > > services: muse: image: codetheweb/muse:PR-640 restart: always volumes: - ./muse:/data environment: - DISCORD_CLIENT_ID= - DISCORD_CLIENT_SECRET= - YOUTUBE_API_KEY= - SPOTIFY_CLIENT_ID= - SPOTIFY_CLIENT_SECRET=
> > 
> > 
> > Looks good, just try it out. I left my `DISCORD_TOKEN` variable in as well tho, I don't know if it's still needed.
> 
> okey where can i find the client id and secret ???

Check the Readme [here](https://github.com/kkendzia/muse/tree/fix/command_permission#running).

You can find the parameters in the OAuth2 Menu:
![image](https://user-images.githubusercontent.com/27204033/166119916-7f934f2f-0609-427f-bb10-9466cc37f362.png)

**Comment by joloman2** [comment]
thank you so verry mutch it works again ty ty ty

**Comment by unixbird** [comment]
So I did this and now im getting the error "could not find settings for guild". I did have to reset my client secret do i just need to reinvite the bot completely again?

**Comment by unixbird** [comment]
> So I did this and now im getting the error "could not find settings for guild". I did have to reset my client secret do i just need to reinvite the bot completely again?

Actually now that i look through logs I'm not even getting a discord link for an invite at all so I'm not sure how to proceed

**Comment by kkendzia** [comment]
Sounds like your database got corrupted... But I cant give you a safe answer about this because i only tested it with an empty directory. You can kick the bot, then you get a invite link - if this doesn't help, delete the `muse` directory where the database is stored. Unfortunatly you would lose all favourites i.e.

**Comment by unixbird** [comment]
> Sounds like your database got corrupted... But I cant give you a safe answer about this because i only tested it with an empty directory. You can kick the bot, then you get a invite link - if this doesn't help, delete the `muse` directory where the database is stored. Unfortunatly you would lose all favourites i.e.

I went ahead and just make an empty directory for it and it produces the same results. I hope making a new bot isn't the solution.

**Comment by unixbird** [comment]
> > Sounds like your database got corrupted... But I cant give you a safe answer about this because i only tested it with an empty directory. You can kick the bot, then you get a invite link - if this doesn't help, delete the `muse` directory where the database is stored. Unfortunatly you would lose all favourites i.e.
> 
> I went ahead and just make an empty directory for it and it produces the same results. I hope making a new bot isn't the solution.

Well made a new bot in the discord dev page and it seemingly works. Already tried resetting everything about the current bot and doesn't work (bot token, client secret) and it gives the same Error: could not find settings for guild which yields no invite link as well.

**Comment by unixbird** [comment]
> > > Sounds like your database got corrupted... But I cant give you a safe answer about this because i only tested it with an empty directory. You can kick the bot, then you get a invite link - if this doesn't help, delete the `muse` directory where the database is stored. Unfortunatly you would lose all favourites i.e.
> > 
> > 
> > I went ahead and just make an empty directory for it and it produces the same results. I hope making a new bot isn't the solution.
> 
> Well made a new bot in the discord dev page and it seemingly works. Already tried resetting everything about the current bot and doesn't work (bot token, client secret) and it gives the same Error: could not find settings for guild which yields no invite link as well.

was able to get the new invite and invite it to a guild however, once i invited it with a clear directory and attempted to set the config roles it yields the same error "could not find settings for guild". It does not seem this PR has fixed the issue for me.

**Comment by Maxximou5** [comment]
Ran into a 400 request issue, due to the [Client Credentials Grant](https://discord.com/developers/docs/topics/oauth2#client-credentials-grant) being limited for team applications.

>     Team applications are limited to the identify and applications.commands.update scope, because teams are not bound to a specific user.

For anyone else running into that issue, obviously you can't use this pr with it set under a team and will need to make a new bot.

**Comment by pierluigizagaria** [comment]
Temporary fixed as simple as that thanks to [PR 640](https://github.com/codetheweb/muse/pull/640).

`podman run -it --name muse -d  -v muse_data:/data --env-file /opt/muse/.env docker.io/codetheweb/muse:33cb3081002abe13bf6c9d56f96f44625429fec6`
or
`docker run -it --name muse -d  -v muse_data:/data --env-file /opt/muse/.env codetheweb/muse:33cb3081002abe13bf6c9d56f96f44625429fec6`

Remember to change the `--env-file` path. 
Should be better than just define all the envs with `-e`

**Comment by unixbird** [comment]
> Temporary fixed as simple as that thanks to [PR 640](https://github.com/codetheweb/muse/pull/640).
> 
> `podman run -it --name muse -d -v muse_data:/data --env-file /opt/muse/.env docker.io/codetheweb/muse:33cb3081002abe13bf6c9d56f96f44625429fec6` or `docker run -it --name muse -d -v muse_data:/data --env-file /opt/muse/.env codetheweb/muse:33cb3081002abe13bf6c9d56f96f44625429fec6`
> 
> Remember to change the `--env-file` path. Should be better than just define all the envs with `-e`

Switching to this image does not yield different results for me. 

still receive the Error: could not find settings for guild  at updatePermissionsForGuild (file:///usr/app/src/utils/update-permissions-for-guild.ts:18:11)

**Comment by kkendzia** [comment]
> Ran into a 400 request issue, due to the [Client Credentials Grant](https://discord.com/developers/docs/topics/oauth2#client-credentials-grant) being limited for team applications.
> 
> > ```
> > Team applications are limited to the identify and applications.commands.update scope, because teams are not bound to a specific user.
> > ```
> 
> For anyone else running into that issue, obviously you can't use this pr with it set under a team and will need to make a new bot.
> 
> You may be able to get away with it by limiting the scope in [PR 640](https://github.com/codetheweb/muse/pull/640/files#diff-2a35bbc57f4f642eedee98f6f99de3282ad81270594478b7a93c485b3a51849aR25) with applications.commands.update.

Yeah, this has something to do with the change of permissions the bot needs. Bots themselves cannot change permissions anymore - also teams are not bound to a specific user. That's why team applications will never again be able to change those permissions.

`Team applications are limited to the identify and applications.commands.update scope, because teams are not bound to a specific user.`

**Comment by kkendzia** [comment]
> > Temporary fixed as simple as that thanks to [PR 640](https://github.com/codetheweb/muse/pull/640).
> > `podman run -it --name muse -d -v muse_data:/data --env-file /opt/muse/.env docker.io/codetheweb/muse:33cb3081002abe13bf6c9d56f96f44625429fec6` or `docker run -it --name muse -d -v muse_data:/data --env-file /opt/muse/.env codetheweb/muse:33cb3081002abe13bf6c9d56f96f44625429fec6`
> > Remember to change the `--env-file` path. Should be better than just define all the envs with `-e`
> 
> Switching to this image does not yield different results for me.
> 
> still receive the Error: could not find settings for guild at updatePermissionsForGuild (file:///usr/app/src/utils/update-permissions-for-guild.ts:18:11)

At least this has nothing to do, as far as I understand, with this issue here. This error occurs when a bot is added to a server, but not present in the database file of the bot (which is locally in the muse directory). 

I added a "fix" for this - to recreate settings of this guild if it isn't readable from the database. This leads to reset settings, but let's the bot start up again.

**Comment by kkendzia** [comment]
Yes in PR640. Huh ok. I can't reproduce this error unfortunately. The only time this occurred to me is when I accidentally deleted the data directory... 
Does this error occurs in docker or when you start it with yarn/npm? If docker whats your env and start command?

**Comment by pierluigizagaria** [comment]
> Temporary fixed

Not really, everytime I play some music the bots stops after some seconds without any log error

**Comment by unixbird** [comment]
> Yes in PR640. Huh ok. I can't reproduce this error unfortunately. The only time this occurred to me is when I accidentally deleted the data directory... Does this error occurs in docker or when you start it with yarn/npm? If docker whats your env and start command?
These are my environment variables.

`environment:
      DISCORD_TOKEN: xxxxx
      DISCORD_CLIENT_ID: xxxxx
      DISCORD_CLIENT_SECRET: xxxxxxxxx
      YOUTUBE_API_KEY: xxxxxxx
      SPOTIFY_CLIENT_ID: xxxxxxx
      SPOTIFY_CLIENT_SECRET: xxxxxxxxx `

and I deploy with docker using:

`docker stack deploy -c muse.yml muse`

 But it seems to do the same thing if i use docker-compose.

**Comment by codetheweb** [comment]
See my reply in the PR fixing this: https://github.com/codetheweb/muse/pull/640#issuecomment-1118066819.

TL;DR waiting until Discord.js merges a PR to merge a fix into master & release.

**Comment by pierluigizagaria** [comment]
Isn't there a temporary workaround? Bot is broken for 10 days

**Comment by pierluigizagaria** [comment]
discordjs branch has been merged to main!
